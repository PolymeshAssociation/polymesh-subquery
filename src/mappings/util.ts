import { decodeAddress } from '@polkadot/keyring';
import { Codec } from '@polkadot/types/types';
import { hexStripPrefix, u8aToHex, u8aToString } from '@polkadot/util';
import { SubstrateEvent, SubstrateExtrinsic } from '@subql/types';
import { Portfolio } from 'polymesh-subql/types/models/Portfolio';
import {
  AssetDocument,
  Compliance,
  Distribution,
  FoundType,
  SecurityIdentifier,
  TransferComplianceExemption,
  TransferManager,
  TransferRestrictionTypeEnum,
} from '../types';
import { Attributes } from './entities/common';

/**
 * @returns a javascript object built using an `iterable` of keys and values.
 * Values are mapped by the map parameter
 */
export const fromEntries = <K extends string | number, V, V2>(
  iterable: Iterable<[K, V]>,
  map: (v: V, i: number, k: K) => V2
): Partial<Record<K, V2>> => {
  const res: Partial<Record<K, V2>> = {};

  let i = 0;
  for (const [k, v] of iterable) {
    res[k] = map(v, i, k);
    i++;
  }
  return res;
};

export const camelToSnakeCase = (str: string): string =>
  str[0].toLowerCase() + str.slice(1).replace(/[A-Z]/g, letter => `_${letter.toLowerCase()}`);

export const snakeToCamelCase = (value: string): string =>
  value
    .toLowerCase()
    .replace(/([-_][a-z])/g, group => group.toUpperCase().replace('-', '').replace('_', ''));

export const capitalizeFirstLetter = (str: string): string => str[0].toUpperCase() + str.slice(1);

export const removeNullChars = (s: string): string => s?.replace(/\0/g, '') || '';

/**
 * @returns the index of the first top level comma in `text` which is a string with nested () and <>.
 * This is meant for rust types, for example "Map<Map<(u8,u32),String>,bool>"" would return 24.
 * @param text
 */
export const findTopLevelCommas = (text: string, exitOnFirst = false): number[] => {
  let nestedLevel = 0;
  const commas = [];
  let i = 0;

  for (const char of text) {
    switch (char) {
      case '(':
      case '{':
      case '<': {
        nestedLevel++;
        break;
      }
      case ')':
      case '}':
      case '>': {
        nestedLevel--;
        break;
      }
      case ',': {
        if (nestedLevel === 1) {
          if (exitOnFirst) {
            return [i];
          } else {
            commas.push(i);
          }
        }
      }
    }
    i++;
  }
  if (commas.length === 0) {
    throw new Error(`No top level comma found in ${text}, it probably isn't a map`);
  }
  return commas;
};

export const getOrDefault = <K, V>(map: Map<K, V>, key: K, getDefault: () => V): V => {
  const v = map.get(key);
  if (v !== undefined) {
    return v;
  } else {
    const def = getDefault();
    map.set(key, def);
    return def;
  }
};

export const serializeTicker = (item: Codec): string => {
  return removeNullChars(u8aToString(item.toU8a()));
};

export const serializeAccount = (item: Codec): string | null => {
  const s = item.toString();

  if (s.trim().length === 0) {
    return null;
  }
  return u8aToHex(decodeAddress(item.toString(), false, item.registry.chainSS58));
};

export const getFirstKeyFromJson = (item: Codec): string => {
  return Object.keys(item.toJSON())[0];
};

export const getFirstValueFromJson = (item: Codec): string => {
  return item.toJSON()[getFirstKeyFromJson(item)];
};

export const getTextValue = (item: Codec): string => {
  return item?.toString().trim().length === 0 ? null : item.toString().trim();
};

export const getNumberValue = (item: Codec): number => {
  return Number(getTextValue(item));
};

export const getDateValue = (item: Codec): Date => {
  return item.toString().trim().length === 0 ? null : new Date(Number(item.toString()));
};

export const getBigIntValue = (item: Codec): bigint => {
  return BigInt(getTextValue(item) || 0);
};

export const getBooleanValue = (item: Codec): boolean => {
  return JSON.parse(getTextValue(item));
};

export const hexToString = (input: string): string => {
  const hex = hexStripPrefix(input);
  let str = '';
  for (let i = 0; i < hex.length; i += 2)
    str += String.fromCharCode(parseInt(hex.substr(i, 2), 16));
  return removeNullChars(str);
};

export const getSigner = (extrinsic: SubstrateExtrinsic): string => {
  return hexStripPrefix(
    u8aToHex(
      decodeAddress(
        extrinsic.extrinsic.signer.toString(),
        false,
        extrinsic.extrinsic.registry.chainSS58
      )
    )
  );
};

/**
 * Parses a raw Asset Document
 */
export const getDocValue = (
  doc: Codec
): Pick<AssetDocument, 'name' | 'link' | 'contentHash' | 'type' | 'filedAt'> => {
  const {
    uri: link,
    content_hash: documentHash,
    name,
    doc_type: type,
    filing_date: filedAt,
  } = JSON.parse(doc.toString());

  const hashType = Object.keys(documentHash)[0];

  return {
    name,
    link,
    contentHash: documentHash[hashType],
    type,
    filedAt,
  };
};

export const getSecurityIdentifiers = (item: Codec): SecurityIdentifier[] => {
  const identifiers = JSON.parse(item.toString());
  return identifiers.map(i => {
    const type = Object.keys(i)[0];
    return {
      type,
      value: i[type],
    };
  });
};

/**
 * Parses a Vec<AssetCompliance>
 */
export const getComplianceRulesValue = (
  requirements: Codec
): Pick<Compliance, 'complianceId' | 'data'>[] => {
  return JSON.parse(JSON.stringify(requirements));
};

/**
 * Parses AssetCompliance
 */
export const getComplianceValue = (
  compliance: Codec
): Pick<Compliance, 'complianceId' | 'data'> => {
  const { id, ...data } = JSON.parse(compliance.toString());
  return {
    complianceId: Number(id),
    data,
  };
};

/**
 * Parses AssetTransferManager
 */
export const getTransferManagerValue = (
  manager: Codec
): Pick<TransferManager, 'type' | 'value'> => {
  const { countTransferManager, percentageTransferManager } = JSON.parse(JSON.stringify(manager));

  if (countTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Count,
      value: Number(countTransferManager),
    };
  }

  if (percentageTransferManager) {
    return {
      type: TransferRestrictionTypeEnum.Percentage,
      value: Number(percentageTransferManager),
    };
  }

  throw new Error('Unknown transfer restriction type found');
};

export const getExemptKeyValue = (
  item: Codec
): Omit<Attributes<TransferComplianceExemption>, 'exemptedEntityId'> => {
  const {
    asset: { ticker },
    op: opType,
    claimType,
  } = JSON.parse(item.toString());
  return {
    assetId: hexToString(ticker),
    opType,
    claimType,
  };
};

export const getExemptionsValue = (exemptions: Codec): string[] => {
  return exemptions.toJSON() as string[];
};

export const logFoundType = (type: string, rawType: string): void => {
  FoundType.create({ id: type, rawType }).save();
};

export const END_OF_TIME = BigInt('253402194600000');

export function addIfNotIncludes<T>(arr: T[], item: T): void {
  if (!arr.includes(item)) {
    arr.push(item);
  }
}

interface MeshPortfolio {
  did: string;
  kind: {
    user: number;
  };
}

const meshPortfolioToPortfolio = ({
  did,
  kind: { user: number },
}: MeshPortfolio): Pick<Portfolio, 'identityId' | 'number'> => ({
  identityId: did,
  number: number || 0,
});

export const getPortfolioValue = (item: Codec): Pick<Portfolio, 'identityId' | 'number'> => {
  const meshPortfolio = JSON.parse(item.toString());
  return meshPortfolioToPortfolio(meshPortfolio);
};

export const getCaIdValue = (item: Codec): Pick<Distribution, 'localId' | 'assetId'> => {
  const { local_id: localId, ticker } = JSON.parse(item.toString());
  return { localId, assetId: hexToString(ticker) };
};

export interface LegDetails {
  from: Pick<Portfolio, 'identityId' | 'number'>;
  to: Pick<Portfolio, 'identityId' | 'number'>;
  ticker: string;
  amount: bigint;
}

export const getLegsValue = (item: Codec): LegDetails[] => {
  const legs = JSON.parse(item.toString());
  return legs.map(({ from: fromPortfolio, to: toPortfolio, asset: ticker, amount }) => ({
    from: meshPortfolioToPortfolio(fromPortfolio),
    to: meshPortfolioToPortfolio(toPortfolio),
    ticker: hexToString(ticker),
    amount: getBigIntValue(amount),
  }));
};

export const getSignerAddress = (event: SubstrateEvent): string => {
  let signer: string;
  if (event.extrinsic) {
    signer = getSigner(event.extrinsic);
  }
  return signer;
};

export const getDistributionValue = (
  item: Codec
): Pick<
  Distribution,
  'portfolioId' | 'currency' | 'perShare' | 'amount' | 'remaining' | 'paymentAt' | 'expiresAt'
> => {
  const { from, currency, per_share, amount, remaining, payment_at, expires_at } = JSON.parse(
    item.toString()
  );
  const { identityId, number } = meshPortfolioToPortfolio(from);
  return {
    portfolioId: `${identityId}/${number}`,
    currency: hexToString(currency),
    perShare: getBigIntValue(per_share),
    amount: getBigIntValue(amount),
    remaining: getBigIntValue(remaining),
    paymentAt: getBigIntValue(payment_at),
    expiresAt: getBigIntValue(expires_at || END_OF_TIME),
  };
};

export const logError = (message: string): void => {
  logger.error(message);
};
