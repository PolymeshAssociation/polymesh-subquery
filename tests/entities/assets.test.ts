import { gql } from '@apollo/client/core';
import { getApolloClient } from '../util';
const { query } = getApolloClient();

describe('assets', () => {
  it('should return all Asset details', async () => {
    const q = {
      query: gql`
        query {
          assets(filter: { ticker: { in: ["4TICKER", "15TICKER", "7TICKER", "11BTICKER1"] } }) {
            nodes {
              nodeId
              id
              ticker
              name
              type
              fundingRound
              isDivisible
              isFrozen
              isUniquenessRequired
              identifiers
              ownerDid: ownerId
              totalSupply
              totalTransfers
              isCompliancePaused
              transferManagers {
                nodes {
                  type
                  value
                  exemptedEntities
                }
              }
              compliance {
                nodes {
                  id: complianceId
                  data
                }
              }
              holders {
                nodes {
                  did: identityId
                  amount
                }
              }
              documents {
                nodes {
                  id: documentId
                  name
                  link
                  contentHash
                  type
                  filedAt
                }
              }
            }
          }
        }
      `,
    };

    const subquery = await query(q);

    expect(subquery?.errors).toBeFalsy();
    expect(subquery?.data).toMatchSnapshot();
  });
});
