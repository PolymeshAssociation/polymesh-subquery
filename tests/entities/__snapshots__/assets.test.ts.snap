// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assets should return all Asset details 1`] = `
Object {
  "assets": Object {
    "__typename": "AssetsConnection",
    "nodes": Array [
      Object {
        "__typename": "Asset",
        "compliance": Object {
          "__typename": "CompliancesConnection",
          "nodes": Array [
            Object {
              "__typename": "Compliance",
              "data": "{\\"sender_conditions\\":[],\\"receiver_conditions\\":[]}",
              "id": 1,
            },
          ],
        },
        "documents": Object {
          "__typename": "AssetDocumentsConnection",
          "nodes": Array [],
        },
        "fundingRound": null,
        "holders": Object {
          "__typename": "AssetHoldersConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetHolder",
              "amount": "999600",
              "did": "0x0500000000000000000000000000000000000000000000000000000000000000",
            },
            Object {
              "__typename": "AssetHolder",
              "amount": "100",
              "did": "0x1c18365371d85ae8662de63bfb0fb90da99b93482e8f043de61ffc83994f8858",
            },
            Object {
              "__typename": "AssetHolder",
              "amount": "100",
              "did": "0x2795441b98ef5a844c7675df8f6030126a297b898ec2d6029dcf81c72d59e553",
            },
            Object {
              "__typename": "AssetHolder",
              "amount": "100",
              "did": "0x2fbb6ef977181d14c7714e134cd5382d42e4214aaf81a2e60b57663946b644c5",
            },
            Object {
              "__typename": "AssetHolder",
              "amount": "100",
              "did": "0x553aa64ffcf49987f124633a835e6363a413cfbd10fb2b77ad8bf17afc7606ff",
            },
          ],
        },
        "id": "11BTICKER1",
        "identifiers": Array [],
        "isCompliancePaused": false,
        "isDivisible": true,
        "isFrozen": false,
        "isUniquenessRequired": false,
        "name": "11BTICKER1",
        "nodeId": "WyJhc3NldHMiLCIxMUJUSUNLRVIxIl0=",
        "ownerDid": "0x0500000000000000000000000000000000000000000000000000000000000000",
        "ticker": "11BTICKER1",
        "totalSupply": "1000000",
        "totalTransfers": "4",
        "transferManagers": Object {
          "__typename": "TransferManagersConnection",
          "nodes": Array [],
        },
        "type": "EquityCommon",
      },
      Object {
        "__typename": "Asset",
        "compliance": Object {
          "__typename": "CompliancesConnection",
          "nodes": Array [
            Object {
              "__typename": "Compliance",
              "data": "{\\"sender_conditions\\":[],\\"receiver_conditions\\":[]}",
              "id": 1,
            },
          ],
        },
        "documents": Object {
          "__typename": "AssetDocumentsConnection",
          "nodes": Array [],
        },
        "fundingRound": null,
        "holders": Object {
          "__typename": "AssetHoldersConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetHolder",
              "amount": "100000",
              "did": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
            },
          ],
        },
        "id": "15TICKER",
        "identifiers": Array [],
        "isCompliancePaused": false,
        "isDivisible": true,
        "isFrozen": false,
        "isUniquenessRequired": false,
        "name": "15TICKER",
        "nodeId": "WyJhc3NldHMiLCIxNVRJQ0tFUiJd",
        "ownerDid": "0x3ece7406def5b26835178d981c4c4a33e9f1b7a79c4432bff1f381c37308ac2f",
        "ticker": "15TICKER",
        "totalSupply": "100000",
        "totalTransfers": "0",
        "transferManagers": Object {
          "__typename": "TransferManagersConnection",
          "nodes": Array [
            Object {
              "__typename": "TransferManager",
              "exemptedEntities": Array [],
              "type": "Count",
              "value": 10,
            },
          ],
        },
        "type": "EquityCommon",
      },
      Object {
        "__typename": "Asset",
        "compliance": Object {
          "__typename": "CompliancesConnection",
          "nodes": Array [],
        },
        "documents": Object {
          "__typename": "AssetDocumentsConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetDocument",
              "contentHash": Object {
                "type": "none",
                "value": null,
              },
              "filedAt": null,
              "id": 0,
              "link": "www.google.com",
              "name": "google",
              "type": null,
            },
          ],
        },
        "fundingRound": null,
        "holders": Object {
          "__typename": "AssetHoldersConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetHolder",
              "amount": "1000000",
              "did": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
            },
          ],
        },
        "id": "4TICKER",
        "identifiers": Array [],
        "isCompliancePaused": false,
        "isDivisible": true,
        "isFrozen": false,
        "isUniquenessRequired": false,
        "name": "4TICKER",
        "nodeId": "WyJhc3NldHMiLCI0VElDS0VSIl0=",
        "ownerDid": "0xc5e2d554233da63d509ee482dbeed0ddec94dc1d0b45ebfdcdc48bd0928222b1",
        "ticker": "4TICKER",
        "totalSupply": "1000000",
        "totalTransfers": "0",
        "transferManagers": Object {
          "__typename": "TransferManagersConnection",
          "nodes": Array [],
        },
        "type": "EquityCommon",
      },
      Object {
        "__typename": "Asset",
        "compliance": Object {
          "__typename": "CompliancesConnection",
          "nodes": Array [
            Object {
              "__typename": "Compliance",
              "data": "{\\"sender_conditions\\":[{\\"condition_type\\":{\\"isPresent\\":{\\"exempted\\":{\\"ticker\\":\\"0x375449434b45520000000000\\"}}},\\"issuers\\":[{\\"issuer\\":\\"0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135\\",\\"trusted_for\\":{\\"any\\":null}}]}],\\"receiver_conditions\\":[{\\"condition_type\\":{\\"isPresent\\":{\\"exempted\\":{\\"ticker\\":\\"0x375449434b45520000000000\\"}}},\\"issuers\\":[{\\"issuer\\":\\"0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135\\",\\"trusted_for\\":{\\"any\\":null}}]}]}",
              "id": 1,
            },
          ],
        },
        "documents": Object {
          "__typename": "AssetDocumentsConnection",
          "nodes": Array [],
        },
        "fundingRound": null,
        "holders": Object {
          "__typename": "AssetHoldersConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetHolder",
              "amount": "1000000",
              "did": "0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135",
            },
          ],
        },
        "id": "7TICKER",
        "identifiers": Array [],
        "isCompliancePaused": false,
        "isDivisible": true,
        "isFrozen": false,
        "isUniquenessRequired": false,
        "name": "7TICKER",
        "nodeId": "WyJhc3NldHMiLCI3VElDS0VSIl0=",
        "ownerDid": "0x0096ea2b9174cbe71c1d3125850b06b2a6e6007cdb3cb33fc371f07154847135",
        "ticker": "7TICKER",
        "totalSupply": "1000000",
        "totalTransfers": "0",
        "transferManagers": Object {
          "__typename": "TransferManagersConnection",
          "nodes": Array [],
        },
        "type": "EquityCommon",
      },
      Object {
        "__typename": "Asset",
        "compliance": Object {
          "__typename": "CompliancesConnection",
          "nodes": Array [
            Object {
              "__typename": "Compliance",
              "data": "{\\"sender_conditions\\":[],\\"receiver_conditions\\":[]}",
              "id": 1,
            },
          ],
        },
        "documents": Object {
          "__typename": "AssetDocumentsConnection",
          "nodes": Array [],
        },
        "fundingRound": "first",
        "holders": Object {
          "__typename": "AssetHoldersConnection",
          "nodes": Array [
            Object {
              "__typename": "AssetHolder",
              "amount": "1000100",
              "did": "0xd7a5d0a364f335632cbbb59bb5615c12286f5b7298f6ddd58b584a6c08734538",
            },
          ],
        },
        "id": "8TICKER",
        "identifiers": Array [],
        "isCompliancePaused": false,
        "isDivisible": true,
        "isFrozen": false,
        "isUniquenessRequired": false,
        "name": "8TICKER",
        "nodeId": "WyJhc3NldHMiLCI4VElDS0VSIl0=",
        "ownerDid": "0xd7a5d0a364f335632cbbb59bb5615c12286f5b7298f6ddd58b584a6c08734538",
        "ticker": "8TICKER",
        "totalSupply": "1000100",
        "totalTransfers": "0",
        "transferManagers": Object {
          "__typename": "TransferManagersConnection",
          "nodes": Array [],
        },
        "type": "EquityCommon",
      },
    ],
  },
}
`;
